src/codelabs/null-safety.md

```diff
@@ -0,0 +1,321 @@
+For example, if you have a variable of
+type `String`, it will always contain a string.
+
+If you want a variable of type `String` to accept any string
+or the value `null`, but nothing else,
+give the variable a nullable type by adding a question mark (`?`) after the
+type name. For example, a variable of type `String?` can contain a string, or
+it can be null.
+
+### Example: Introducing non-nullable types
+
+The variable `a` below is declared as an `int`. Try changing the value in the
+assignment to 3 or 145. Anything but null!
+
+<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/non_nullable_types.dart" replace="/145/null/g"?>
+```dart:run-dartpad:ga_id-nonnullable_type:null_safety-true
+void main() {
+  int a;
+  a = null;
+  print('a is $a.');
+}
+```
+
+### Example: Nullable types
+
+What if you need a variable that *can* hold a null value?  Try changing the
+type of `a` so that `a` can be either null or an int:
+
+<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/nullable_types.dart" replace="/int\?/int/g"?>
+```dart:run-dartpad:ga_id-nonnullable_type:null_safety-true
+void main() {
+  int a;
+  a = null;
+  print('a is $a.');
+}
+```
+
+### Example: More nullable types!
+
+Type parameters for generics can also be nullable or non-nullable. Try using
+question marks to correct the type declarations of `aNullableListOfStrings` and
+`aListOfNullableStrings`:
+
+<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/more_nullable_types.dart" replace="/String\?/String/g"?>
+```dart:run-dartpad:ga_id-nonnullable_type:null_safety-true
+void main() {
+  List<String> aListofStrings = ['one', 'two', 'three'];
+  List<String> aNullableListOfStrings = [];
+  List<String> aListofNullableStrings = ['one', null, 'three'];
+
+  print('aListofStrings is $aListofStrings.');
+  print('aNullableListOfStrings is $aNullableListOfStrings.');
+  print('aListofNullableStrings is $aListofNullableStrings.');
+}
+```
+
+## Using ?, !, and late
+
+Null safety introduces new operators and keywords for dealing with null values,
+`?`, `!`, and `late`. Here's how you use them:
+
+-   If you're creating a variable that can hold a null value, you can use a
+    `?` to inform Dart of the variable’s nullability.
+-   If you're sure that an expression with a nullable type isn't null, you can 
+    add a postfix exclamation mark (`!`) to make Dart treat it as non-nullable.
+-   If you want to initalize a non-nullable variable to a
+    non-null value before it’s used, insert `late` before the
+    variable’s type.
+
+### Example: The assertion operator (!)
+
+If you'd like to assign a nullable expression to a variable that's
+non-nullable, you can use the assertion operator (the exclamation point: `!`).
+By adding `!` just after the expression, you tell Dart that the value won't be
+null, and that it's safe to assign it to a non-nullable variable. If you're
+wrong, Dart throws an exception.
```

Comment:
s/an exception/an exception at run-time/ to make it very clear.

Reply:
Maybe also add: "This makes the `!` operator unsafe, and you should only use it when you are very sure that the expression isn't null."